ipv8/test/gossip/test_community.py

```diff
@@ -0,0 +1,144 @@
+from twisted.internet import reactor
+from twisted.internet.task import deferLater
+
+from ...gossip.community import GossipOverlay, GossipRule, IGossipOverlayListener
+from ...peer import Peer
+from ...test.base import TestBase
+from ...test.mocking.ipv8 import MockIPv8
+from ...test.util import twisted_wrapper
+
+
+class MockGossipOverlayListener(IGossipOverlayListener):
+
+    def __init__(self, node):
+        """
+        Initialize a Mock Gossip Listener
+
+        :param node: a node which must have a Gossip Overlay
+        """
+        self.node = node
+
+    def on_gossip(self, public_key, message):
+        self.node.overlay.store(public_key, message)
+
+
+class TestGossipOverlay(TestBase):
+
+    def create_node(self, *args, **kwargs):
+        return MockIPv8(u"curve25519", self.overlay_class, *args, **kwargs)
+
+    def setUp(self):
+        super(TestGossipOverlay, self).setUp()
+        self.initialize(GossipOverlay, 3)
+
+        overlay_prefix = self.nodes[0].overlay.prefix
+
+        # Introduce the neighbors for the Gossip service
+        for node in self.nodes:
+            for other in self.nodes:
+                if not other == node:
+                    private_peer = other.my_peer
+                    public_peer = Peer(private_peer.public_key, private_peer.address)
+                    node.network.add_verified_peer(public_peer)
+                    node.network.discover_services(public_peer, [overlay_prefix])
+
+    def remove_neighborhood(self, peer):
+        """
+        Removes the neighborhood of a peer
+
+        :param peer: the peer whose neighborhood should be removed
+        :return: None
+        """
+        for node in self.nodes:
+            if not peer == node:
+                private_peer = node.my_peer
+                public_peer = Peer(private_peer.public_key, private_peer.address)
+                peer.network.remove_peer(public_peer)
+
+    @twisted_wrapper(2)
+    def test_gossip_rule_vote_success(self):
+        """
+        Test the voting scheme for changing the GossipRule for a particular peer
+
+        :return: None
+        """
+        target_public_key = self.nodes[0].overlay.my_peer.public_key.key_to_bin()
+
+        self.nodes[1].overlay.set_rule(target_public_key, GossipRule.SUPPRESS)
+        self.nodes[2].overlay.set_rule(target_public_key, GossipRule.SUPPRESS)
+
+        yield deferLater(reactor, 1, lambda: None)
```

Comment:
We have `self.sleep(1)` for this, but requiring a sleep is a sign of a bad threading model in the first place.

Reply:
The reason for the wait here is due to the `LoopingCall` in the `GossipOverlay`, which fires every `0.5` seconds. I've added a parameter (called `loop_interval`) for this interval which defaults to the aforementioned value. It is possible to avoid using a `deferLater` by simply reducing it to say `0.005` seconds. I can modify the creation of peers such that the `loop_interval` has a small value.

Reply:
It is notoriously very difficult to debug looping calls. To solve this, we usually cancel all pending tasks in the overlay of each node/peer (https://github.com/Tribler/py-ipv8/blob/master/ipv8/test/messaging/anonymization/test_community.py#L35) and then invoke the periodic callback manually.

This ensures that you do not depend on sleeps or get very nasty race conditions.

Reply:
Ah, I see. This makes more sense. I'll opt for this solution then.